src/Pate/Verification/AbstractDomain.hs

```diff
@@ -0,0 +1,506 @@
+
+data AbstractDomainBody sym arch where
+  AbstractDomainBody ::
+    { absDomEq :: PED.EquivalenceDomain sym arch
+      -- ^ specifies which locations are equal between the original vs. patched programs
+    , absDomVals :: PPa.PatchPair (AbstractDomainVals sym arch)
+      -- ^ specifies independent constraints on the values for the original and patched programs
+    } -> AbstractDomainBody sym arch
+
+
+
+-- | Defining the known range for a given location. Currently this is either a constant
+-- value or an unconstrained value.
+data AbsRange (tp :: W4.BaseType) where
+  AbsIntConstant :: Integer -> AbsRange W4.BaseIntegerType
+  AbsBVConstant :: 1 W4.<= w => W4.NatRepr w -> BV.BV w -> AbsRange (W4.BaseBVType w)
+  AbsBoolConstant :: Bool -> AbsRange W4.BaseBoolType
+  AbsUnconstrained :: W4.BaseTypeRepr tp -> AbsRange tp
+
+groundToAbsRange :: W4.BaseTypeRepr tp -> W4G.GroundValue tp -> AbsRange tp
+groundToAbsRange repr v = case repr of
+  W4.BaseIntegerRepr -> AbsIntConstant v
+  W4.BaseBVRepr w -> AbsBVConstant w v
+  W4.BaseBoolRepr -> AbsBoolConstant v
+  _ -> AbsUnconstrained repr
+
+
+absRangeRepr :: AbsRange tp -> W4.BaseTypeRepr tp
+absRangeRepr r = case r of
+  AbsIntConstant{} -> W4.BaseIntegerRepr
+  AbsBVConstant w _ -> W4.BaseBVRepr w
+  AbsBoolConstant{} -> W4.BaseBoolRepr
+  AbsUnconstrained repr -> repr
+
+
+instance TestEquality AbsRange where
+  testEquality r1 r2 = case (r1, r2) of
+    (AbsIntConstant i1, AbsIntConstant i2) | i1 == i2 -> Just Refl
+    (AbsBVConstant w1 bv1, AbsBVConstant w2 bv2)
+      | Just Refl <- testEquality w1 w2, bv1 == bv2 -> Just Refl
+    (AbsBoolConstant b1, AbsBoolConstant b2) | b1 == b2 -> Just Refl
+    (AbsUnconstrained repr1, AbsUnconstrained repr2) -> testEquality repr1 repr2
+    _ -> Nothing
+
+instance Eq (AbsRange tp) where
+  r1 == r2 = case testEquality r1 r2 of
+    Just Refl -> True
+    Nothing -> False
+
+-- | Find an 'AbsRange' that represents the intersection of two ranges.
+-- In the case where the ranges don't intersect (e.g. they represent two different
+-- constant values), the resulting range is unconstrained.
+combineAbsRanges :: AbsRange tp -> AbsRange tp -> AbsRange tp
+combineAbsRanges r1 r2 = if r1 == r2 then r1 else AbsUnconstrained (absRangeRepr r1)
+
+-- | An abstract value keyed on the macaw type, which defines an independent
+-- 'AbsRange' for each component base expression (i.e. for a pointer there is a separate
+-- range for the region and offset).
+data MacawAbstractValue sym (tp :: MT.Type) where
+  MacawAbstractValue ::
+      { macawAbsVal :: Ctx.Assignment AbsRange (PSR.CrucBaseTypes (MS.ToCrucibleType tp))
+      }
+      -> MacawAbstractValue sym tp
+  deriving Eq
+
+-- | An abstract value for a pointer, with a separate 'AbsRange' for the region and offset.
+data MemAbstractValue sym w where
+  MemAbstractValue :: (MacawAbstractValue sym (MT.BVType (8 W4.* w))) -> MemAbstractValue sym w
+
+
+-- | Mapping from locations to abstract values.
```

Comment:
At the moment we aren't reasoning at all about this, no. This is establishing constraints about values sitting in memory, but doesn't have any information about the pointers themselves. I think this would make sense as a separate kind of abstract domain, which specifically tracks information about symbolic pointers.

Reply:
Can we just add a note about that and some potential issues? We don't need to fix it right now